README.md
# ArangoDB Exporter for Prometheus

This exporter exposes the statistics provided by a specific ArangoDB instance
in a format compatible with prometheus.

## Usage

To use the ArangoDB Exporter, run the following:

```bash
arangodb-exporter \
    --arangodb.endpoint=http://<your-database-host>:8529 \
    --arangodb.jwtsecret=<your-jwt-secret> \
    --ssl.keyfile=<your-optional-ssl-keyfile>
```

This results in an ArangoDB Exporter exposing all statistics of
the ArangoDB server (running at `http://<your-database-host>:8529`)
at `http://<your-host-ip>:9101/metrics`.

## Running in Docker

To run the ArangoDB Exporter in docker, use an image such as
[`arangodb/arangodb-exporter:0.1.3`](https://hub.docker.com/r/arangodb/arangodb-exporter/).

## Configuring Prometheus

There are several ways to configure Prometheus to fetch metrics from the ArangoDB Exporter.

Below you're find a sample Prometheus configuration file that can be used to fetch
metrics from an ArangoDB exporter listening on localhost port 9101 (without TLS).

```yaml
global:
  scrape_interval:     15s
scrape_configs:
- job_name: arangodb
  static_configs:
  - targets: ['localhost:9101']
```

For more info on configuring Prometheus go to [its configuration documentation](https://prometheus.io/docs/prometheus/latest/configuration/configuration).

If you're using the [Prometheus Operator](https://github.com/coreos/prometheus-operator)
in Kubernetes, you need to create an additional `Service` and a `ServiceMonitor` resource
like this:

```yaml
kind: Service
apiVersion: v1
metadata:
  name: arangodb-exporters-service
  labels:
    app: arangodb-exporter
spec:
  selector:
    app: arangodb-exporter
  ports:
  - name: metrics
    port: 9101

---

apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: arangodb-exporter
  namespace: monitoring
  labels:
    team: frontend
    prometheus: kube-prometheus
spec:
  namespaceSelector:
    matchNames:
    - default
  selector:
    matchLabels:
      app: arangodb-exporter
  endpoints:
  - port: metrics
    scheme: https
    tlsConfig:
      insecureSkipVerify: true
```

Note 1: that the typical deployment on the Prometheus operator is done in
a namespace called `monitoring`. Make sure to match the `namespace`
of the `ServiceMonitor` to match that.

Note 2: that the `Prometheus` custom resource has a field called `serviceMonitorSelector`.
Make sure that the `matchLabels` selector in there matches the labels of
your `ServiceMonitor`.

## Building

To build this project, you need Go 1.8 or higher and Docker installed.
Then run:

```bash
DOCKERNAMESPACE=<your docker hub account name> make
```

CHANGELOG
0.1.6 (2019-05-09)
------------------

* Move to go 1.12 and modules

0.1.5 (2019-01-22)
------------------

* Trim whitespaces when token is read from file.

0.1.4 (2019-01-17)
------------------

* Added --arangodb.jwt-file option so that a file containing the jwt can be used by the exporter to authenticate to the database.
  This is *not* the jwt secret. Please refere to the arangodb-starter documentation for how to create such a token.


exporter_test.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"fmt"
	_ "net/http/pprof"
	"strings"
	"testing"

	"github.com/prometheus/client_golang/prometheus"
)

// TestMetric tests the result of metricKey & newMetric for various inputs.
func TestMetric(t *testing.T) {
	g1 := StatisticGroup{
		Group:       "g1",
		Name:        "g1-name",
		Description: "Something g1",
	}
	tests := []struct {
		Group               StatisticGroup
		Figure              StatisticFigure
		Postfix             string
		KeyResult           string
		CollectionPostfixes []string
	}{
		{g1, StatisticFigure{"g1", "f1", "f1-name", "descr-f1", FigureTypeAccumulated, "", nil}, "_pf", "g1-name_f1-name_pf", []string{""}},
		{g1, StatisticFigure{"g1", "f1", "f1-name", "descr-f1", FigureTypeAccumulated, "tick", nil}, "_pf", "g1-name_f1-name_pf_tick", []string{""}},
		{g1, StatisticFigure{"g1", "f2", "f2-name", "descr-f2", FigureTypeDistribution, "", []float64{0.1, 0.2}}, "_pf", "g1-name_f2-name_pf", []string{"_sum", "_count", "_bucket"}},
	}

	for i, test := range tests {
		result := metricKey(test.Group, test.Figure, test.Postfix)
		if result != test.KeyResult {
			t.Errorf("metricKey for test %d failed: got '%s', expected '%s'", i, result, test.KeyResult)
		}
		colls := newMetric(test.Group, test.Figure)
		if len(colls) != len(test.CollectionPostfixes) {
			t.Errorf("newMetric for test %d returns unexpected #collectors: got %d, expected %d", i, len(colls), len(test.CollectionPostfixes))
		} else {
			for ci, c := range colls {
				descrChan := make(chan *prometheus.Desc, 1)
				c.Describe(descrChan)
				d := <-descrChan
				result := d.String()
				expectedPostfix := test.CollectionPostfixes[ci]
				if !strings.Contains(result, fmt.Sprintf("%s\", help", expectedPostfix)) {
					t.Errorf("newMetric for test %d returns collector %d with wrong expectation. got '%s', expected it to contain '%s'", i, ci, result, expectedPostfix)
				}
			}
		}
	}
}


jwt.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	jg "github.com/dgrijalva/jwt-go"
)

const (
	issArangod = "arangodb"
)

// CreateArangodJwtAuthorizationHeader calculates a JWT authorization header, for authorization
// of a request to an arangod server, based on the given secret.
// If the secret is empty, nothing is done.
func CreateArangodJwtAuthorizationHeader(jwt string) (string, error) {
	return "bearer " + jwt, nil
}

// CreateArangodJWT creates a superuser arangod jwt
func CreateArangodJWT(jwtSecret string) (string, error) {
	if jwtSecret == "" {
		return "", nil
	}
	// Create a new token object, specifying signing method and the claims
	// you would like it to contain.
	token := jg.NewWithClaims(jg.SigningMethodHS256, jg.MapClaims{
		"iss":       issArangod,
		"server_id": "exporter",
	})

	// Sign and get the complete encoded token as a string using the secret
	signedToken, err := token.SignedString([]byte(jwtSecret))
	if err != nil {
		return "", maskAny(err)
	}

	return signedToken, nil
}


server.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"crypto/tls"
	"net/http"
	_ "net/http/pprof"
	"time"

	certificates "github.com/arangodb-helper/go-certificates"
)

// ServerConfig settings for the Server
type ServerConfig struct {
	Address    string // Address to listen on
	TLSKeyfile string // Keyfile containing TLS certificate
}

// Server is the HTTPS server for the operator.
type Server struct {
	httpServer *http.Server
}

// NewServer creates a new server, fetching/preparing a TLS certificate.
func NewServer(handler http.Handler, cfg ServerConfig) (*Server, error) {
	httpServer := &http.Server{
		Addr:              cfg.Address,
		Handler:           handler,
		ReadTimeout:       time.Second * 30,
		ReadHeaderTimeout: time.Second * 15,
		WriteTimeout:      time.Second * 30,
		TLSNextProto:      make(map[string]func(*http.Server, *tls.Conn, http.Handler)),
	}

	if cfg.TLSKeyfile != "" {
		tlsConfig, err := createTLSConfig(cfg.TLSKeyfile)
		if err != nil {
			return nil, maskAny(err)
		}
		tlsConfig.BuildNameToCertificate()
		httpServer.TLSConfig = tlsConfig
	}

	return &Server{
		httpServer: httpServer,
	}, nil
}

// Run the server until the program stops.
func (s *Server) Run() error {
	if s.httpServer.TLSConfig != nil {
		if err := s.httpServer.ListenAndServeTLS("", ""); err != nil && err != http.ErrServerClosed {
			return maskAny(err)
		}
	} else {
		if err := s.httpServer.ListenAndServe(); err != nil && err != http.ErrServerClosed {
			return maskAny(err)
		}
	}
	return nil
}

// createTLSConfig creates a TLS config from the given keyfile.
func createTLSConfig(keyfile string) (*tls.Config, error) {
	cert, err := certificates.LoadKeyFile(keyfile)
	if err != nil {
		return nil, maskAny(err)
	}
	return &tls.Config{
		Certificates: []tls.Certificate{cert},
	}, nil
}


arangodb.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"context"
	_ "net/http/pprof"

	driver "github.com/arangodb/go-driver"
)

// StatisticsDescription is the JSON representation of the result of an _admin/statistics-description call.
type StatisticsDescription struct {
	Groups  []StatisticGroup  `json:"groups"`
	Figures []StatisticFigure `json:"figures"`
}

// StatisticGroup describes a group of statistics.
type StatisticGroup struct {
	Group       string `json:"group"`
	Name        string `json:"name"`
	Description string `json:"description"`
}

// StatisticFigure describes a single statistic.
type StatisticFigure struct {
	Group       string     `json:"group"`
	Identifier  string     `json:"identifier"`
	Name        string     `json:"name"`
	Description string     `json:"description"`
	Type        FigureType `json:"type"`
	Units       string     `json:"unit,omitempty"`
	Cuts        []float64  `json:"cuts,omitempty"`
}

// FigureType is a strongly typed type of statistic.
type FigureType string

const (
	FigureTypeAccumulated  FigureType = "accumulated"
	FigureTypeCurrent      FigureType = "current"
	FigureTypeDistribution FigureType = "distribution"
)

// Statistics is a stringly typed map containing statistic values
type Statistics map[string]interface{}

// GetGroup returns the statistics for the given group.
// If not found, nil is returned
func (s Statistics) GetGroup(group string) Statistics {
	entry, ok := s[group]
	if !ok {
		return nil
	}
	result, ok := entry.(map[string]interface{})
	if !ok {
		return nil
	}
	return result
}

// GetFloat returns a float value for a statistic with given key.
// If not found, or the value could not be converted to a float, false is returned.
func (s Statistics) GetFloat(key string) (float64, bool) {
	raw, ok := s[key]
	if !ok {
		return 0.0, false
	}
	if f, ok := raw.(float64); ok {
		return f, true
	}
	if f, ok := raw.(int64); ok {
		return float64(f), true
	}
	return 0.0, false
}

// GetCounts returns an int64 array value for a statistic with given key.
// If not found, or the value could not be converted to an int64 array, false is returned.
func (s Statistics) GetCounts(key string) ([]int64, bool) {
	raw, ok := s[key]
	if !ok {
		return nil, false
	}
	if rawSlice, ok := raw.([]interface{}); ok {
		result := make([]int64, len(rawSlice))
		for i, x := range rawSlice {
			if v, ok := x.(int64); ok {
				result[i] = v
			} else if v, ok := x.(int); ok {
				result[i] = int64(v)
			} else if v, ok := x.(float64); ok {
				result[i] = int64(v)
			} else {
				return nil, false
			}
		}
		return result, true
	}
	return nil, false
}

// GetInt returns an int value for a statistic with given key.
// If not found, or the value could not be converted to an int, false is returned.
func (s Statistics) GetInt(key string) (int64, bool) {
	raw, ok := s[key]
	if !ok {
		return 0.0, false
	}
	if i, ok := raw.(int64); ok {
		return i, true
	}
	if i, ok := raw.(int); ok {
		return int64(i), true
	}
	return 0, false
}

// GetStatistics requests the statistics values from the given connection.
func GetStatistics(ctx context.Context, conn driver.Connection) (Statistics, error) {
	req, err := conn.NewRequest("GET", "_admin/statistics")
	if err != nil {
		return nil, maskAny(err)
	}
	resp, err := conn.Do(ctx, req)
	if err != nil {
		return nil, maskAny(err)
	}
	if err := resp.CheckStatus(200); err != nil {
		return nil, maskAny(err)
	}
	var result Statistics
	if err := resp.ParseBody("", &result); err != nil {
		return nil, maskAny(err)
	}
	return result, nil
}

// GetStatisticsDescription requests the statistics description from the given connection.
func GetStatisticsDescription(ctx context.Context, conn driver.Connection) (StatisticsDescription, error) {
	req, err := conn.NewRequest("GET", "_admin/statistics-description")
	if err != nil {
		return StatisticsDescription{}, maskAny(err)
	}
	resp, err := conn.Do(ctx, req)
	if err != nil {
		return StatisticsDescription{}, maskAny(err)
	}
	if err := resp.CheckStatus(200); err != nil {
		return StatisticsDescription{}, maskAny(err)
	}
	var result StatisticsDescription
	if err := resp.ParseBody("", &result); err != nil {
		return StatisticsDescription{}, maskAny(err)
	}
	return result, nil
}


exporter.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"context"
	"crypto/tls"
	"fmt"
	_ "net/http/pprof"
	"strings"
	"sync"
	"time"

	driver "github.com/arangodb/go-driver"
	driver_http "github.com/arangodb/go-driver/http"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/log"
)

const (
	namespace = "arangodb" // For Prometheus metrics.

)

// metricKey returns a key into the map of metrics for the given figure & group.
func metricKey(group StatisticGroup, figure StatisticFigure, postfix string) string {
	result := strings.Replace(strings.ToLower(group.Name+"_"+figure.Name), " ", "_", -1)
	if postfix != "" {
		result = result + postfix
	}
	if figure.Units != "" {
		return result + "_" + strings.ToLower(figure.Units)
	}
	return result
}

// newMetric creates one or more metrics for the given figure & group.
func newMetric(group StatisticGroup, figure StatisticFigure) []prometheus.Collector {
	switch figure.Type {
	case FigureTypeDistribution:
		return []prometheus.Collector{
			// _sum
			prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      metricKey(group, figure, "_sum"),
				Help:      figure.Description,
			}),
			// _count
			prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      metricKey(group, figure, "_count"),
				Help:      figure.Description,
			}),
			// _bucket
			prometheus.NewGaugeVec(prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      metricKey(group, figure, "_bucket"),
				Help:      figure.Description,
			}, []string{"le"}),
		}
	default:
		return []prometheus.Collector{
			prometheus.NewGauge(prometheus.GaugeOpts{
				Namespace: namespace,
				Name:      metricKey(group, figure, ""),
				Help:      figure.Description,
			}),
		}
	}
}

// Exporter collects ArangoDB statistics from the given endpoint and exports them using
// the prometheus metrics package.
type Exporter struct {
	conn    driver.Connection
	timeout time.Duration
	mutex   sync.RWMutex

	metrics                     map[string][]prometheus.Collector
	up                          prometheus.Gauge
	totalScrapes, failedScrapes prometheus.Counter
}

// NewExporter returns an initialized Exporter.
func NewExporter(arangodbEndpoint, jwt string, sslVerify bool, timeout time.Duration) (*Exporter, error) {
	connCfg := driver_http.ConnectionConfig{
		Endpoints: []string{arangodbEndpoint},
	}
	if !sslVerify {
		connCfg.TLSConfig = &tls.Config{InsecureSkipVerify: true}
	}
	conn, err := driver_http.NewConnection(connCfg)
	if err != nil {
		return nil, maskAny(err)
	}
	if jwt != "" {
		hdr, err := CreateArangodJwtAuthorizationHeader(jwt)
		if err != nil {
			return nil, maskAny(err)
		}
		auth := driver.RawAuthentication(hdr)
		conn, err = conn.SetAuthentication(auth)
		if err != nil {
			return nil, maskAny(err)
		}
	}

	return &Exporter{
		conn:    conn,
		timeout: timeout,
		up: prometheus.NewGauge(prometheus.GaugeOpts{
			Namespace: namespace,
			Name:      "up",
			Help:      "Was the last scrape of ArangoDB successful.",
		}),
		totalScrapes: prometheus.NewCounter(prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "exporter_total_scrapes",
			Help:      "Current total ArangoDB scrapes.",
		}),
		failedScrapes: prometheus.NewCounter(prometheus.CounterOpts{
			Namespace: namespace,
			Name:      "exporter_failed_scrapes",
			Help:      "Number of failed ArangoDB scrapes",
		}),
		metrics: make(map[string][]prometheus.Collector),
	}, nil
}

// Describe describes all the metrics ever exported by the HAProxy exporter. It
// implements prometheus.Collector.
func (e *Exporter) Describe(ch chan<- *prometheus.Desc) {
	for _, ms := range e.metrics {
		for _, m := range ms {
			m.Describe(ch)
		}
	}
	ch <- e.up.Desc()
	ch <- e.totalScrapes.Desc()
	ch <- e.failedScrapes.Desc()
}

// Collect fetches the stats from ArangoDB statistics and delivers them
// as Prometheus metrics. It implements prometheus.Collector.
func (e *Exporter) Collect(ch chan<- prometheus.Metric) {
	e.mutex.Lock() // To protect metrics from concurrent collects.
	defer e.mutex.Unlock()

	ctx, cancel := context.WithTimeout(context.Background(), e.timeout)
	defer cancel()
	e.resetMetrics()
	e.scrape(ctx)

	ch <- e.up
	ch <- e.totalScrapes
	ch <- e.failedScrapes
	e.collectMetrics(ch)
}

// scrape performs a single query of all statistics.
func (e *Exporter) scrape(ctx context.Context) {
	e.totalScrapes.Inc()

	// Gather descriptions
	descr, err := GetStatisticsDescription(ctx, e.conn)
	if err != nil {
		e.up.Set(0)
		log.Errorf("Failed to fetch statistic descriptions: %v", err)
		return
	}

	// Collect statistics
	stats, err := GetStatistics(ctx, e.conn)
	if err != nil {
		e.up.Set(0)
		log.Errorf("Failed to fetch statistics: %v", err)
		return
	}

	// Mark ArangoDB as up.
	e.up.Set(1)

	// Now parse the statistics & put them in the correct metrics
	groups := make(map[string]StatisticGroup)
	for _, g := range descr.Groups {
		groups[g.Group] = g
	}
	for _, f := range descr.Figures {
		group, found := groups[f.Group]
		if !found {
			// Skip figure with unknown group
			continue
		}
		groupStats := stats.GetGroup(f.Group)
		if groupStats == nil {
			// Skip no group is found in the statistics
		}
		key := metricKey(group, f, "")
		ms, found := e.metrics[key]
		if !found {
			ms = newMetric(group, f)
			e.metrics[key] = ms
		}
		switch f.Type {
		case FigureTypeCurrent, FigureTypeAccumulated:
			if value, ok := groupStats.GetFloat(f.Identifier); ok {
				gauge := ms[0].(prometheus.Gauge)
				gauge.Set(value)
			}
		case FigureTypeDistribution:
			distStats := groupStats.GetGroup(f.Identifier)
			if distStats != nil {
				// _sum comes first
				if sum, ok := distStats.GetFloat("sum"); ok {
					gauge := ms[0].(prometheus.Gauge)
					gauge.Set(sum)
				}
				// _count comes second
				if sum, ok := distStats.GetFloat("count"); ok {
					gauge := ms[1].(prometheus.Gauge)
					gauge.Set(sum)
				}
				// _bucket comes third
				if counts, ok := distStats.GetCounts("counts"); ok {
					gaugeVec := ms[2].(*prometheus.GaugeVec)
					cummulative := int64(0)
					for i, v := range counts {
						var leValue string
						if i < len(f.Cuts) {
							leValue = fmt.Sprintf("%v", f.Cuts[i])
						} else {
							leValue = "+Inf"
						}
						gaugeVec.WithLabelValues(leValue).Set(float64(cummulative + v))
						cummulative += v
					}
				}
			}
		}
	}
}

type resetter interface {
	Reset()
}

func (e *Exporter) resetMetrics() {
	for _, ms := range e.metrics {
		for _, m := range ms {
			if r, ok := m.(resetter); ok {
				r.Reset()
			}
		}
	}
}

func (e *Exporter) collectMetrics(metrics chan<- prometheus.Metric) {
	for _, ms := range e.metrics {
		for _, m := range ms {
			if c, ok := m.(prometheus.Collector); ok {
				c.Collect(metrics)
			}
		}
	}
}


tools/release/release.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"crypto/sha256"
	"encoding/hex"
	"flag"
	"fmt"
	"io/ioutil"
	"log"
	"os"
	"os/exec"
	"path/filepath"
	"strings"

	"github.com/coreos/go-semver/semver"
)

var (
	versionFile string // Full path of VERSION file
	releaseType string // What type of release to create (major|minor|patch)
	ghRelease   string // Full path of github-release tool
	ghUser      string // Github account name to create release in
	ghRepo      string // Github repository name to create release in
	binFolder   string // Folder containing binaries

	binaries = map[string]string{
		"arangodb-exporter-darwin-amd64":      "darwin/amd64/arangodb-exporter",
		"arangodb-exporter-linux-amd64":       "linux/amd64/arangodb-exporter",
		"arangodb-exporter-linux-arm":         "linux/arm/arangodb-exporter",
		"arangodb-exporter-linux-arm64":       "linux/arm64/arangodb-exporter",
		"arangodb-exporter-windows-amd64.exe": "windows/amd64/arangodb-exporter.exe",
	}
)

func init() {
	flag.StringVar(&versionFile, "versionfile", "./VERSION", "Path of the VERSION file")
	flag.StringVar(&releaseType, "type", "patch", "Type of release to build (major|minor|patch)")
	flag.StringVar(&ghRelease, "github-release", "bin/github-release", "Full path of github-release tool")
	flag.StringVar(&ghUser, "github-user", "arangodb-helper", "Github account name to create release in")
	flag.StringVar(&ghRepo, "github-repo", "arangodb-exporter", "Github repository name to create release in")
	flag.StringVar(&binFolder, "bin-folder", "./bin", "Folder containing binaries")
}

func main() {
	flag.Parse()
	ensureGithubToken()
	checkCleanRepo()
	version := bumpVersion(releaseType)
	envVars := map[string]string{
		"DOCKERNAMESPACE": "arangodb",
		"DOCKERTAG":       version,
	}
	//make("clean", envVars)
	make("build", envVars)
	createSHA256Sums()
	make("docker", envVars)
	gitTag(version)
	githubCreateRelease(version)
	bumpVersion("devel")
}

// ensureGithubToken makes sure the GITHUB_TOKEN envvar is set.
func ensureGithubToken() {
	token := os.Getenv("GITHUB_TOKEN")
	if token == "" {
		p := filepath.Join(os.Getenv("HOME"), ".arangodb/github-token")
		if raw, err := ioutil.ReadFile(p); err != nil {
			log.Fatalf("Failed to release '%s': %v", p, err)
		} else {
			token = strings.TrimSpace(string(raw))
			os.Setenv("GITHUB_TOKEN", token)
		}
	}
}

func checkCleanRepo() {
	output, err := exec.Command("git", "status", "--porcelain").Output()
	if err != nil {
		log.Fatalf("Failed to check git status: %v\n", err)
	}
	if strings.TrimSpace(string(output)) != "" {
		log.Fatal("Repository has uncommitted changes\n")
	}
}

func make(target string, envVars map[string]string) {
	if err := run("make", []string{target}, envVars); err != nil {
		log.Fatalf("Failed to make %s: %v\n", target, err)
	}
}

func bumpVersion(action string) string {
	contents, err := ioutil.ReadFile(versionFile)
	if err != nil {
		log.Fatalf("Cannot read '%s': %v\n", versionFile, err)
	}
	version := semver.New(strings.TrimSpace(string(contents)))

	switch action {
	case "patch":
		version.BumpPatch()
	case "minor":
		version.BumpMinor()
	case "major":
		version.BumpMajor()
	case "devel":
		version.Metadata = "git"
	}
	contents = []byte(version.String())

	if err := ioutil.WriteFile(versionFile, contents, 0755); err != nil {
		log.Fatalf("Cannot write '%s': %v\n", versionFile, err)
	}

	gitCommitAll(fmt.Sprintf("Updated to %s", version))
	log.Printf("Updated '%s' to '%s'\n", versionFile, string(contents))

	return version.String()
}

func gitCommitAll(message string) {
	args := []string{
		"commit",
		"--all",
		"-m", message,
	}
	if err := run("git", args, nil); err != nil {
		log.Fatalf("Failed to commit: %v\n", err)
	}
	if err := run("git", []string{"push"}, nil); err != nil {
		log.Fatalf("Failed to push commit: %v\n", err)
	}
}

func gitTag(version string) {
	if err := run("git", []string{"tag", version}, nil); err != nil {
		log.Fatalf("Failed to tag: %v\n", err)
	}
	if err := run("git", []string{"push", "--tags"}, nil); err != nil {
		log.Fatalf("Failed to push tags: %v\n", err)
	}
}

func createSHA256Sums() {
	sums := []string{}
	for name, p := range binaries {
		blob, err := ioutil.ReadFile(filepath.Join(binFolder, p))
		if err != nil {
			log.Fatalf("Failed to read binary '%s': %#v\n", name, err)
		}
		bytes := sha256.Sum256(blob)
		sha := hex.EncodeToString(bytes[:])
		sums = append(sums, sha+"  "+name)
	}
	sumsPath := filepath.Join(binFolder, "SHA256SUMS")
	if err := ioutil.WriteFile(sumsPath, []byte(strings.Join(sums, "\n")+"\n"), 0644); err != nil {
		log.Fatalf("Failed to write '%s': %#v\n", sumsPath, err)
	}
}

func githubCreateRelease(version string) {
	// Create draft release
	args := []string{
		"release",
		"--user", ghUser,
		"--repo", ghRepo,
		"--tag", version,
		"--draft",
	}
	if err := run(ghRelease, args, nil); err != nil {
		log.Fatalf("Failed to create github release: %v\n", err)
	}
	// Upload binaries
	assets := map[string]string{
		"SHA256SUMS": "SHA256SUMS",
	}
	for k, v := range binaries {
		assets[k] = v
	}
	for name, file := range assets {
		args := []string{
			"upload",
			"--user", ghUser,
			"--repo", ghRepo,
			"--tag", version,
			"--name", name,
			"--file", filepath.Join(binFolder, file),
		}
		if err := run(ghRelease, args, nil); err != nil {
			log.Fatalf("Failed to upload asset '%s': %v\n", name, err)
		}
	}
	// Finalize release
	args = []string{
		"edit",
		"--user", ghUser,
		"--repo", ghRepo,
		"--tag", version,
	}
	if err := run(ghRelease, args, nil); err != nil {
		log.Fatalf("Failed to finalize github release: %v\n", err)
	}
}

func run(cmd string, args []string, envVars map[string]string) error {
	c := exec.Command(cmd, args...)
	if envVars != nil {
		c.Env = append(c.Env, os.Environ()...)
		for k, v := range envVars {
			c.Env = append(c.Env, fmt.Sprintf("%s=%s", k, v))
		}
	}
	c.Stdout = os.Stdout
	c.Stderr = os.Stderr
	return c.Run()
}


main.go
//
// DISCLAIMER
//
// Copyright 2018 ArangoDB GmbH, Cologne, Germany
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
// http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// Copyright holder is ArangoDB GmbH, Cologne, Germany
//
// Author Ewout Prangsma
//

package main

import (
	"fmt"
	"io/ioutil"
	"net/http"
	_ "net/http/pprof"
	"strings"
	"time"

	"github.com/pkg/errors"
	"github.com/prometheus/client_golang/prometheus"
	"github.com/prometheus/common/log"
	"github.com/prometheus/common/version"
	"github.com/spf13/cobra"
)

var (
	projectVersion = "dev"
	projectBuild   = "dev"
	maskAny        = errors.WithStack

	cmdMain = &cobra.Command{
		Use: "arangodb-exporter",
		Run: cmdMainRun,
	}

	serverOptions   ServerConfig
	arangodbOptions struct {
		endpoint  string
		jwtSecret string
		jwtFile   string
		timeout   time.Duration
	}
)

func init() {
	f := cmdMain.Flags()

	f.StringVar(&serverOptions.Address, "server.address", ":9101", "Address the exporter will listen on (IP:port)")
	f.StringVar(&serverOptions.TLSKeyfile, "ssl.keyfile", "", "File containing TLS certificate used for the metrics server. Format equal to ArangoDB keyfiles")

	f.StringVar(&arangodbOptions.endpoint, "arangodb.endpoint", "http://127.0.0.1:8529", "Endpoint used to reach the ArangoDB server")
	f.StringVar(&arangodbOptions.jwtSecret, "arangodb.jwtsecret", "", "JWT Secret used for authentication with ArangoDB server")
	f.StringVar(&arangodbOptions.jwtFile, "arangodb.jwt-file", "", "File containing the JWT for authentication with ArangoDB server")
	f.DurationVar(&arangodbOptions.timeout, "arangodb.timeout", time.Second*15, "Timeout of statistics requests for ArangoDB")

	f.MarkDeprecated("arangodb.jwtsecret", "please use --arangodb.jwt-file instead")
}

func main() {
	cmdMain.Execute()
}

func cmdMainRun(cmd *cobra.Command, args []string) {
	log.Infoln(fmt.Sprintf("Starting arangodb-exporter %s, build %s", projectVersion, projectBuild))

	var token string
	if arangodbOptions.jwtFile != "" {
		data, err := ioutil.ReadFile(arangodbOptions.jwtFile)
		if err != nil {
			log.Fatal(err)
		}
		token = strings.TrimSpace(string(data))
	} else if arangodbOptions.jwtSecret != "" {
		var err error
		token, err = CreateArangodJWT(arangodbOptions.jwtSecret)
		if err != nil {
			log.Fatal(err)
		}
	}

	exporter, err := NewExporter(arangodbOptions.endpoint, token, false, arangodbOptions.timeout)
	if err != nil {
		log.Fatal(err)
	}
	prometheus.MustRegister(exporter)
	version.Version = projectVersion
	version.Revision = projectBuild
	prometheus.MustRegister(version.NewCollector("arangodb_exporter"))

	log.Infoln("Listening on", serverOptions.Address)

	mux := http.NewServeMux()
	mux.Handle("/metrics", prometheus.Handler())
	mux.HandleFunc("/", func(w http.ResponseWriter, r *http.Request) {
		w.Write([]byte(`<html>
             <head><title>ArangoDB Exporter</title></head>
             <body>
             <h1>ArangoDB Exporter</h1>
             <p><a href='/metrics'>Metrics</a></p>
             </body>
             </html>`))
	})

	server, err := NewServer(mux, serverOptions)
	if err != nil {
		log.Fatal(err)
	}
	log.Fatal(server.Run())
}
